### PR TITLE
adjusted M3 docs

### PR DESCRIPTION
**Pull request by enikao**
* Moved serialization and M3 docs to separate directories
* Reordered sections in M3 doc
* Updated pre-defined ids and keys in M3 / builtins
* Added builtin `INamed` and use it in M3
* Added `Concept.partition` property
* Added reference definitions of M3 and builtins as JSON
* Updated serialization format class diagram

### CODE REVIEW

File: lioncore/metametamodel/lioncore.json [hunk 1]
```diff
@@ -0,0 +1,2196 @@
...
+  "serializationFormatVersion": "1",
+  "languages": [
+    {
+      "key": "LIonCore-M3",
+      "version": "1"
+    }
+  ],
+  "nodes": [
+    {
+      "id": "-id-LIonCore-M3",
```
Comment (joswarmer) [inline lioncore/metametamodel/lioncore.json:11]: All id values start with a minus, is this really the intention? Or is something missing?

File: lioncore/metametamodel/lioncore.json [hunk 2]
```diff
@@ -0,0 +1,2196 @@
...
+      "concept": {
+        "language": "LIonCore-M3",
+        "version": "1",
+        "key": "Language"
+      },
+      "properties": [
+        {
+          "property": {
+            "language": "LIonCore-M3",
```
Comment (joswarmer) [inline lioncore/metametamodel/lioncore.json:20]: Did we not agree on using underscores instead of minusses? Or is that at some other place?

Comment (joswarmer) [reply lioncore/metametamodel/lioncore.json:20]: I checked, and in previous versions we user underscores. I prefer underscores, because they are allowed in almost every environment within identifiers, while minusses are not.

File: lioncore/metametamodel/lioncore.json [hunk 3]
```diff
@@ -0,0 +1,2196 @@
...
+            "version": "1",
+            "key": "LIonCore-builtins-INamed-name"
+          },
+          "value": "LIonCore.M3"
```
Comment (joswarmer) [inline lioncore/metametamodel/lioncore.json:24]: And here we use a dot. I am confused as to where (and why) we use dot, minus and underscore.

File: lioncore/metametamodel/lioncore.json (hunk 1 above)
Comment (enikao) [reply lioncore/metametamodel/lioncore.json:11]: I deliberately used different _ids_ and _keys_, so it's clear everywhere which one we refer to.
In principle the _id_ should be arbitrary (as long as they resolve). As I wrote this file by hand, I used kind of meaningful ids.

File: lioncore/metametamodel/lioncore.json (hunk 2 above)
Comment (enikao) [reply lioncore/metametamodel/lioncore.json:20]: In #91 we decided to change _keys_ to use minuses (rationale in the issue).

File: lioncore/metametamodel/lioncore.json (hunk 3 above)
Comment (enikao) [reply lioncore/metametamodel/lioncore.json:24]: This is the name of the language. It has no technical meaning, so could also be "LIonWeb LIonCore M3 Metamodel" or anything else.
As naming is hard, I chose this one without any imagination. And made it different from _key_ and _id_ for aforementioned reason.

Comment (joswarmer) [reply lioncore/metametamodel/lioncore.json:24]: I agree naming is hard indeed !

File: lioncore/metametamodel/lioncore.json [hunk 4]
```diff
@@ -0,0 +1,2196 @@
...
+            "version": "1",
+            "key": "Language-elements"
+          },
+          "children": [
+            "-id-Concept",
+            "-id-ConceptInterface",
+            "-id-Containment",
+            "-id-DataType",
+            "-id-Enumeration",
+            "-id-EnumerationLiteral",
```
Comment (joswarmer) [inline lioncore/metametamodel/lioncore.json:56]: EnumerationLiteral is a containment within Enumeration, not within Language (see question on slack)

File: lioncore/metametamodel/lioncore.json [hunk 5]
```diff
@@ -0,0 +1,2196 @@
...
+            "-id-Feature",
```
Comment (joswarmer) [inline lioncore/metametamodel/lioncore.json:57]: Feature is contained in FeaturesContainer, nit in Language.

File: lioncore/metametamodel/lioncore.json [hunk 6]
```diff
@@ -0,0 +1,2180 @@
...
+        {
+          "reference": {
+            "language": "LIonCore-M3",
+            "version": "1",
+            "key": "Concept-implements"
+          },
+          "targets": []
+        }
+      ],
+      "parent": "LIonCore-M3"
```
Comment (ftomassetti) [inline lioncore/metametamodel/lioncore.json:2042]: This should be "-id-LIonCore-M3", I think

File: lioncore/metametamodel/lioncore.json [hunk 7]
```diff
@@ -0,0 +1,2180 @@
...
+        {
+          "property": {
+            "language": "LIonCore-M3",
+            "version": "1",
+            "key": "LIonCore-builtins-INamed-name"
+          },
+          "value": "NamespacedEntity"
+        },
+        {
+          "property": {
```
Comment (ftomassetti) [inline lioncore/metametamodel/lioncore.json:1747]: Is Key a property of NamespacedEntity? I was not expecting this

In the Java implementation I have an interface called HasKey, but we discussed the fact that it should be just an implementation detail, so we do not have a common point where to define the key.

I think that NamespacedEntity could not play this role, as the Language has a key but it is not a NamespacedEntity.

I think that we have two options:
- either we use two key properties (one declared under Language and one under LanguageElement)
- or we introduce a common ancestore for Language and LanguageElement which defines the key

Comment (enikao) [reply lioncore/metametamodel/lioncore.json:1747]: Currently option 1 is implemented here. Option 2 would be `HasKey`. As you say, we decided against `HasKey`, which would leave Option 1.
Conceptually, I agree `key` is a bit out of place within `NamespacedEntity` -- `key` is a language thing, namespaces could be used outside this context.
We could rediscuss this tomorrow?

File: lioncore/metametamodel/lioncore.json (hunk 4 above)
Comment (enikao) [reply lioncore/metametamodel/lioncore.json:56]: as discussed on slack: correct as-is.

File: lioncore/metametamodel/lioncore.json (hunk 5 above)
Comment (enikao) [reply lioncore/metametamodel/lioncore.json:57]: as discussed on slack: correct as-is.

File: lioncore/metametamodel/builtins.json [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+      "id": "LIonCore-builtins",
+      "concept": {
+        "language": "LIonCore-M3",
+        "version": "1",
+        "key": "Language"
+      },
+      "properties": [
+        {
+          "property": {
+            "language": "LIonCore-M3",
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply lioncore/metametamodel/builtins.json]: Yes, indeed

File: lioncore/metametamodel/lioncore.json (hunk 7 above)
Comment (ftomassetti) [reply lioncore/metametamodel/lioncore.json:1747]: Sure, makes sense

File: lioncore/metametamodel/builtins.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (enikao) [reply lioncore/metametamodel/builtins.json]: already applied (-:

File: lioncore/metametamodel/builtins.json [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+          },
+          "value": "String"
+        },
+        {
+          "property": {
+            "language": "LIonCore-M3",
+            "version": "1",
+            "key": "IKeyed-key"
+          },
+          "value": "LIonCore-builtins-String"
```
Comment (ftomassetti) [inline lioncore/metametamodel/builtins.json:94]: For M3 ids are different from keys, while for the builtins they match. For consistency I would make them different, following the same schema used in M3

File: lioncore/metametamodel/lioncore.json [hunk 10]
```diff
@@ -0,0 +1,2066 @@
...
+        {
+          "reference": {
+            "language": "LIonCore-M3",
+            "version": "1",
+            "key": "Property-type"
+          },
+          "targets": [
+            {
+              "resolveInfo": "Boolean",
+              "reference": "LIonCore-M3-Boolean"
```
Comment (ftomassetti) [inline lioncore/metametamodel/lioncore.json:205]: I think this should be `LIonCore-builtins-Boolean`

File: lioncore/metametamodel/lioncore.json [hunk 11]
```diff
@@ -0,0 +1,2066 @@
...
+        {
+          "reference": {
+            "language": "LIonCore-M3",
+            "version": "1",
+            "key": "Concept-implements"
+          },
+          "targets": [
+            {
+              "resolveInfo": "IKeyed",
+              "reference": "IKeyed"
```
Comment (ftomassetti) [inline lioncore/metametamodel/lioncore.json:1411]: I think this should be `-id-IKeyed`

File: lioncore/metametamodel/lioncore.json [hunk 12]
```diff
@@ -0,0 +1,2066 @@
...
+        {
+          "reference": {
+            "language": "LIonCore-M3",
+            "version": "1",
+            "key": "Property-type"
+          },
+          "targets": [
+            {
+              "resolveInfo": "String",
+              "reference": "LIonCore-M3-String"
```
Comment (ftomassetti) [inline lioncore/metametamodel/lioncore.json:1580]: I think this should be `LIonCore-builtins-String`

File: lioncore/metametamodel/lioncore.json [hunk 13]
```diff
@@ -0,0 +1,2066 @@
...
+            "-id-IKeyed-key"
+          ]
+        }
+      ],
+      "references": [
+        {
+          "reference": {
+            "language": "LIonCore-M3",
+            "version": "1",
+            "key": "Concept-extends"
```
Comment (ftomassetti) [inline lioncore/metametamodel/lioncore.json:1705]: IKeyed is an interface, therefore it is using `ConceptInterface-extends`

File: lioncore/metametamodel/lioncore.json [hunk 14]
```diff
@@ -0,0 +1,2066 @@
...
+      "children": [
+        {
+          "containment": {
+            "language": "LIonCore-M3",
+            "version": "1",
+            "key": "Classifier-features"
+          },
+          "children": [
+            "-id-Concept-abstract",
+            "-id-Concept-extends",
```
Comment (ftomassetti) [inline lioncore/metametamodel/lioncore.json:131]: It should also have the partition feature

File: lioncore/metametamodel/lioncore.json [hunk 15]
```diff
@@ -0,0 +1,2066 @@
...
+              "resolveInfo": "Language",
+              "reference": "-id-Language"
+            }
+          ]
+        }
+      ],
+      "parent": "-id-Language"
+    },
+    {
+      "id": "-id-Language-elements",
```
Comment (ftomassetti) [inline lioncore/metametamodel/lioncore.json:1478]: Should this become entities?

File: lioncore/metametamodel/builtins.json (hunk 9 above)
Comment (enikao) [reply lioncore/metametamodel/builtins.json:94]: This is something I'm not sure about, see https://lionweb-org.github.io/organization/lioncore/metametamodel/metametamodel.html#_ids_of_built_in_elements

What do you think?

File: lioncore/metametamodel/lioncore.json (hunk 10 above)
Comment (enikao) [reply lioncore/metametamodel/lioncore.json:205]: fixed

File: lioncore/metametamodel/lioncore.json (hunk 11 above)
Comment (enikao) [reply lioncore/metametamodel/lioncore.json:1411]: fixed

File: lioncore/metametamodel/lioncore.json (hunk 12 above)
Comment (enikao) [reply lioncore/metametamodel/lioncore.json:1580]: fixed

File: lioncore/metametamodel/lioncore.json (hunk 13 above)
Comment (enikao) [reply lioncore/metametamodel/lioncore.json:1705]: fixed (removed `Concept-extends`, renamed `Concept-implements` to `ConceptInterface-extends`)

File: lioncore/metametamodel/lioncore.json (hunk 14 above)
Comment (enikao) [reply lioncore/metametamodel/lioncore.json:131]: fixed

File: lioncore/metametamodel/builtins.json (hunk 9 above)
Comment (ftomassetti) [reply lioncore/metametamodel/builtins.json:94]: I understand that both keys and IDs must be stable, but I do not understand why they need to be identical. However I have just a very tiny preference for them being difficult for consistency with M3 and to catch bugs more easily, but definitely not a strong opinion about this

Comment (enikao) [reply lioncore/metametamodel/builtins.json:94]: My line of thought: Node ids must only be consistent within one _context_. A context might be one json file, or one repository. Once we move from one context to another (e.g. importing a json file into a repository), all ids might change. That's ok, as long as they change consistently.

(If we later wanted to import another json file that refers to the first one, our repository has to keep the ids of the first one around, but that import issue is not relevant here.)

Keys must be consistent _accross contexts_. Otherwise we always have to ship the M2 along with the M1. (We already decided that.)
This also applies to the keys in `builtins`.

The `builtins` model is special in the sense that we _never_ ship it between contexts -- we always assume it's there.
Thus, we also need stable ids in `builtins`, otherwise we cannot refer to them. But we have to refer to them in every M2.

So yes, we can keep keys and ids different, as long as both are stable.

For `lioncore.json`, the ids (in form `-id-xxx`) don't really matter, as they are gone after import to our repository. We never want to refer to these nodes from the same meta-level (nobody can add/remove things at M3 level). References from other meta-levels use keys.
If we wanted to refer to M3 elements as regular models (e.g. in a documentation language: "Please create a new ->M3.Concept"), these elements are "just" regular nodes, so we refer to them by which id they have in the local context.

For `builtins.json`, the ids _do_ matter. That's why I think the consistency between `lioncore` and `builtins` is not that relevant.

I agree debugging is easier if id differs from key. On the other hand, implementations have to keep two sets of "magic" strings around.

Long story short: I have a slight preference for keeping key == id for `builtins`, but also not a strong opinion.

Comment (enikao) [reply lioncore/metametamodel/builtins.json:94]: Decision for now: Keep the same. Revisit once we have experience or discover issues.
